cmd/gcp-controller-manager/csr_signer.go

```diff
@@ -67,35 +71,52 @@ func newGKESigner(ctx *controllerContext) (*gkeSigner, error) {
 	}, nil
 }
 
+// handle is called by the generic Kubernetes certificate controller.
 func (s *gkeSigner) handle(csr *capi.CertificateSigningRequest) error {
+	_, _, err := s.handleInternal(csr)
+	return err
+}
+
+func (s *gkeSigner) handleInternal(csr *capi.CertificateSigningRequest) (processed bool, out *capi.CertificateSigningRequest, err error) {
 	if !certificates.IsCertificateRequestApproved(csr) {
-		return nil
+		return false, nil, nil
+	}
+
+	// Ignore CSRs that are not addressed to the default signer.
+	if csr.Spec.SignerName != nil &&
+		*csr.Spec.SignerName != certsv1b1.KubeAPIServerClientSignerName &&
+		*csr.Spec.SignerName != certsv1b1.KubeAPIServerClientKubeletSignerName &&
+		*csr.Spec.SignerName != certsv1b1.KubeletServingSignerName &&
+		*csr.Spec.SignerName != certsv1b1.LegacyUnknownSignerName {
+		return false, nil, nil
 	}
+
 	klog.Infof("gkeSigner triggered for %q", csr.Name)
 
 	recordMetric := csrmetrics.SigningStartRecorder(authFlowLabelNone)
 	x509cr, err := certutil.ParseCSR(csr.Spec.Request)
 	if err != nil {
 		recordMetric(csrmetrics.SigningStatusParseError)
-		return fmt.Errorf("unable to parse csr %q: %v", csr.Name, err)
+		return true, nil, fmt.Errorf("unable to parse csr %q: %v", csr.Name, err)
 	}
 	recordMetric = csrmetrics.SigningStartRecorder(s.metricLabel(csr, x509cr))
 	csr, err = s.sign(csr)
 	if err != nil {
 		recordMetric(csrmetrics.SigningStatusSignError)
-		return fmt.Errorf("error auto signing csr: %v", err)
+		return true, nil, fmt.Errorf("error auto signing csr: %v", err)
 	}
 	updateRecordMetric := csrmetrics.OutboundRPCStartRecorder("k8s.CertificateSigningRequests.updateStatus")
-	_, err = s.ctx.client.CertificatesV1beta1().CertificateSigningRequests().UpdateStatus(context.TODO(), csr, metav1.UpdateOptions{})
+	newCSR, err := s.ctx.client.CertificatesV1beta1().CertificateSigningRequests().UpdateStatus(context.TODO(), csr, metav1.UpdateOptions{})
 	if err != nil {
 		updateRecordMetric(csrmetrics.OutboundRPCStatusError)
 		recordMetric(csrmetrics.SigningStatusUpdateError)
-		return fmt.Errorf("error updating signature for csr: %v", err)
+		return true, nil, fmt.Errorf("error updating signature for csr: %v", err)
 	}
+
 	updateRecordMetric(csrmetrics.OutboundRPCStatusOK)
 	klog.Infof("CSR %q signed", csr.Name)
 	recordMetric(csrmetrics.SigningStatusSigned)
-	return nil
+	return true, csr, nil
```

Comment:
should this return `newCSR`?

Reply:
Done (by renaming newCSR to csr).

I also added a .DeepCopy() into sign(), since from what I remember it's not OK to modify an object that came out of a shared informer.

Reply:
> I also added a .DeepCopy() into sign(), since from what I remember it's not OK to modify an object that came out of a shared informer.

Good thought, though in this case, the certificate controller copies before calling handle() to protect against this:

https://github.com/kubernetes/cloud-provider-gcp/blob/c7b4749cfddac2532f0ef743965297180e945d78/vendor/k8s.io/kubernetes/pkg/controller/certificates/certificate_controller.go#L194-L205

Reply:
Oh, neat.  I will undo it.

Reply:
Done